### ISSUE TITLE
Talkback does not read the expected text

### ISSUE THREAD

**Issue by JishnuGoyal** [oppia/oppia-android#4739]
**Describe the bug**

On the enter pin page on a user profile, the talkback directly reads the enter pin text. Expectation: It should start from Hi, (user name), Please enter your pin.

**To Reproduce**
1. Go to a user profile which has a pin with talkback turned on
2. Talkback directly is focussed on text input field and reads "Pin verification".

**Environment**
- Device/emulator being used: <Vivo V25 pro>
- Android or SDK version (e.g. Android 5 or SDK 21): Android 12
- App version (you can get this through system app settings or via the admin controls menu in-app): 0.10-beta-5e64fae55e

**Comment by BenHenning** [comment]
This seems like an a11y nice-to-have, not a broken flow, so we don't need to block the release.

**Comment by Vishwajith-Shettigar** [comment]
can i work on this issue ?

**Comment by adhiamboperes** [comment]
> 


Hi @Vishwajith-Shettigar, could you work on getting your first PR approved, then you can start on this other issue? Thanks!

**Comment by Vishwajith-Shettigar** [comment]
> > 
> 
> Hi @Vishwajith-Shettigar, could you work on getting your first PR approved, then you can start on this other issue? Thanks!

yeah okay

**Comment by adhiamboperes** [comment]
@Vishwajith-Shettigar, could you please begin work on this issue?

**Comment by Vishwajith-Shettigar** [comment]
> @Vishwajith-Shettigar, could you please begin work on this issue?

Sure thank you

**Comment by Vishwajith-Shettigar** [comment]
![Screenshot_2023-09-08-19-36-02-48_943a62cb4c6fb83e010e1c2e82766a17](https://github.com/oppia/oppia-android/assets/76042077/405d8b7d-8a13-4c48-a6e9-ae8fad4237ca)
@adhiamboperes  Can you  please confirm me issue is addressing the same page !?

**Comment by Vishwajith-Shettigar** [comment]
![Screenshot_2023-09-08-19-36-02-48_943a62cb4c6fb83e010e1c2e82766a17](https://github.com/oppia/oppia-android/assets/76042077/c1aef53a-69c3-4eb1-9ba0-8364420cafcc)
@adhiamboperes Can you please confirm me issue is addressing the same page in screenshot

**Comment by MohitGupta121** [comment]
@Vishwajith-Shettigar correct this is same.

**Comment by adhiamboperes** [comment]
> @Vishwajith-Shettigar correct this is same.

Thanks @MohitGupta121. @Vishwajith-Shettigar,  please let us know if you have any questions about the solution

**Comment by Vishwajith-Shettigar** [comment]
@adhiamboperes Talkback reads activity label every time we open the activity so what could be the solution?, Im thinking to make that activity label empty. Please suggest me any good solution.

**Comment by Vishwajith-Shettigar** [comment]
@adhiamboperes  We are expecting when activity opens talk back should read Hi, (user name), Please enter your pin. then after talk back will focus on Edit text. Am I going right ?

**Comment by Vishwajith-Shettigar** [comment]
> @Vishwajith-Shettigar correct this is same.

thank you

**Comment by adhiamboperes** [comment]
> @adhiamboperes We are expecting when activity opens talk back should read Hi, (user name), Please enter your pin. then after talk back will focus on Edit text. Am I going right ?


The user greeting textview needs a content description to be read, same to the enter pin textview

**Comment by Vishwajith-Shettigar** [comment]
> > @adhiamboperes We are expecting when activity opens talk back should read Hi, (user name), Please enter your pin. then after talk back will focus on Edit text. Am I going right ?
> 
> The user greeting textview needs a content description to be read, same to the enter pin textview
Yes but, in what sequence, issue is stating Talk back should "Hi, (user name), Please enter your pin" instead of "pin verification".
But currently it starts with "pin verification" and focus on edit text. Talk back first reads activity label which is "pin verification"
So my solution would be Changing activity label to " Hi, (user name), Please enter your pin" dynamically. If I don't do like that and it always says "pin verification" and if I prevent focus from edit text it focus on navigationicon in toolbar.

**Comment by Vishwajith-Shettigar** [comment]
> > @adhiamboperes We are expecting when activity opens talk back should read Hi, (user name), Please enter your pin. then after talk back will focus on Edit text. Am I going right ?
> 
> The user greeting textview needs a content description to be read, same to the enter pin textview

Talk back reads activity label first that is "pin verification" now what should i do to mute that, am i allowed to remove activity label ?

**Comment by adhiamboperes** [comment]
> > The user greeting textview needs a content description to be read, same to the enter pin textview
> 
> Talk back reads activity label first that is "pin verification" now what should i do to mute that, am i allowed to remove activity label ?


What we need is for talkback to read the texts on the screen in the order in which they appear. The way to do this is to make sure the textviews all have corresponding content descriptions. The activity labels shoud not be modified. To clear any confusion caused by the issue description, please just make sure what we have on the screen is what is actually being read.

**Comment by Vishwajith-Shettigar** [comment]
> > > The user greeting textview needs a content description to be read, same to the enter pin textview
> > 
> > 
> > Talk back reads activity label first that is "pin verification" now what should i do to mute that, am i allowed to remove activity label ?
> 
> What we need is for talkback to read the texts on the screen in the order in which they appear. The way to do this is to make sure the textviews all have corresponding content descriptions. The activity labels shoud not be modified. To clear any confusion caused by the issue description, please just make sure what we have on the screen is what is actually being read.

all i just need is to give content description right ? or something else ?

**Comment by adhiamboperes** [comment]
Please start there and confirm that the solution works as expected. Afterwards you can also check the corresponding test files and add tests for the new content descriptions. You can find the pattern for testing from existing tests on the same.

**Comment by Vishwajith-Shettigar** [comment]
@adhiamboperes i set content description for user greeting textview and enter pin textview, but the issue is  talback directly focus on edit text when we enter the activity.

**Comment by adhiamboperes** [comment]
> @adhiamboperes i set content description for user greeting textview and enter pin textview, but the issue is talback directly focus on edit text when we enter the activity.

There is an xml attribute `android:importantForAccessibility="yes"` that tells talkback a view is important. Please research it a bit and then try it.

**Comment by Vishwajith-Shettigar** [comment]
> > @adhiamboperes i set content description for user greeting textview and enter pin textview, but the issue is talback directly focus on edit text when we enter the activity.
> 
> There is an xml attribute `android:importantForAccessibility="yes"` that tells talkback a view is important. Please research it a bit and then try it.

I have tried that, if i make android:importantForAccessibility="no" for edit text that works but edit text wont be able to read by talkback even if user touch it

**Comment by Vishwajith-Shettigar** [comment]
> > > @adhiamboperes i set content description for user greeting textview and enter pin textview, but the issue is talback directly focus on edit text when we enter the activity.
> > 
> > 
> > There is an xml attribute `android:importantForAccessibility="yes"` that tells talkback a view is important. Please research it a bit and then try it.
> 
> I have tried that, if i make android:importantForAccessibility="no" for edit text that works but edit text wont be able to read by talkback even if user touch it

and i have tried android:importantForAccessibility="yes" for textviews but there is no difference

**Comment by adhiamboperes** [comment]
Can you please research this and let me know what you come up with?

**Comment by Vishwajith-Shettigar** [comment]
Okay i will try

**Comment by Vishwajith-Shettigar** [comment]
I came up with a solution, talkback is directly focusing on edit text because of edit text auto focus  by ,making edittext focusable="false" and dynamically setting focusable="true" when user click on edit text, this might solve.

**Comment by adhiamboperes** [comment]
That sounds like a great idea. Please go ahead with the implementation.

**Comment by Vishwajith-Shettigar** [comment]
> That sounds like a great idea. Please go ahead with the implementation.

@adhiamboperes  here is PR #5152

**Comment by adhiamboperes** [comment]
I have reviewed the talkback experience on this screen---it sounds to me like the text being read is sufficient for the user to know what is required, and I don't think there is need to modify it at this point. Here is a recording. @seanlip , @BenHenning, please review.
 
https://github.com/oppia/oppia-android/assets/59600948/f38570b0-c52b-4199-bca9-4887075759bc

**Comment by seanlip** [comment]
@adhiamboperes I'm not sure what the expected behaviour for a11y is when there is autofocus enabled -- @BenHenning might know better here so it might be worth asking him directly.

One thought I had -- will they know that there's a "Forgot PIN?" option on the screen, if we don't enable the whole screen to be read at the outset? I feel like, based on this, I lean towards thinking it should be changed along the lines that @JishnuGoyal proposed, if only for consistency with other screens (and thinking about the case where we might add more stuff to this screen in future).

Just my 2c.

**Comment by seanlip** [comment]
Also to clarify -- my previous comment was meant to describe what I think is the "ideal" case, but I think that fixing this is still probably fairly low impact and that it might be better to work on higher-impact issues instead.

(@adhiamboperes just a note, we probably shouldn't really be putting "good first labels" on low-impact projects in general, if there are other things that would be more important to do. Better to break down the higher-impact issues if possible so that they're approachable.)

**Comment by adhiamboperes** [comment]
Thanks @seanlip, I'll keep that in mind.
